### PR TITLE
Tweak highlight to detect corpus + misc. tweaks

### PR DESCRIPTION
**Pull request by SKevo18**
Changes:
- split highlight tokens to literals and regex, to (possibly) speed things up a little;
- added a lot of new tokens for highlight, to (hopefully) not slow things down a little;
- verbose now prints list of undetected files, for debugging;
- counter for files with suspect lines;
- show percentage of `statistic` out of `total files` in bottom stat table;
- add a flag to scan all valid UTF-8 files, regardless of file extension (possibly redundant);

Example:

```
Corediff completed scanning 1066 files in folder/test
 - Files with unrecognized lines      : 208    19.51%
 - Files with suspect lines           : 208    19.51%
 - Files with only recognized lines   : 0     0.00%
 - Files with custom code             : 0     0.00%
 - Files without text                 : 858    80.49%
```

Todo:
- [x] Are there actually 858 non-text files in backend corpus? Also tried with new `-t` (valid UTF-8 only) flag

Ref.: https://www.pivotaltracker.com/n/projects/2320827/stories/187212227

### CODE REVIEW

File: corediff.go [hunk 1]
```diff
@@ -244,14 +257,23 @@ func main() {
...
-			logInfo(" - Files without code              :", stats.filesNoCode)
+			logInfo(" - Files with unrecognized lines      :", boldred(stats.filesWithChanges), grey(fmt.Sprintf("%8.2f%%", stats.percentage(stats.filesWithChanges))))
+			logInfo(" - Files with suspect lines           :", warn(stats.filesWithSuspectLines), grey(fmt.Sprintf("%8.2f%%", stats.percentage(stats.filesWithSuspectLines))))
+			logInfo(" - Files with only recognized lines   :", green(stats.filesWithoutChanges), grey(fmt.Sprintf("%8.2f%%", stats.percentage(stats.filesWithoutChanges))))
+			logInfo(" - Files with custom code             :", stats.filesCustomCode, grey(fmt.Sprintf("%8.2f%%", stats.percentage(stats.filesCustomCode))))
+			logInfo(" - Files without", without, "                :", stats.filesNoCode, grey(fmt.Sprintf("%8.2f%%", stats.percentage(stats.filesNoCode))))
+			logVerbose("Undetected paths:")
+			for _, p := range stats.undetectedPaths {
+				logVerbose("  ", p)
+			}
```
Comment (SKevo18) [inline corediff.go:276]: Alternatively:

```suggestion
            if args.Verbose {
			  logInfo("Undetected paths:")
			  for _, p := range stats.undetectedPaths {
				  logInfo("  ", p)
			  }
			}
```

File: highlight.go [hunk 2]
```diff
@@ -1,25 +1,95 @@
...
+		[]byte(`chgrp($`),
+		[]byte(`unlink(`),
+		[]byte(`rmdir(`),
+		[]byte(`mkdir(`),
+		[]byte(`stream_get_meta_data`),
+
+		[]byte(`GLOBALS`),
+
+		[]byte(`$obirninja`),
+		[]byte(`$pass`),
```
Comment (SKevo18) [inline highlight.go:80]:
```suggestion
```

Possibly too many FPs (it's supposed to match password variables in reverse shells, etc... - should probably make it more specific then)

Comment (danslo) [reply highlight.go:80]: IMO better to highlight a FP than to not highlight a TP.

Let's be more greedy :)

Comment (danslo) [reply highlight.go:80]: IMO better to highlight a FP than to not highlight a TP.

Let's be more greedy :)

File: corediff.go (hunk 1 above)
Comment (gwillem) [reply corediff.go:276]: logverbose seems good to me!

**Comment by gwillem** [comment]
> Are there actually 858 non-text files in backend corpus?

```
❯ find sigs/corpus/backend -type f | xargs file | grep -v text | wc -l
      84
```

File: corediff.go [hunk 3]
```diff
@@ -186,9 +194,14 @@ func addPath(root string, db hashDB, args *baseArgs) {
 			return nil
 		}
 
-		if !hasValidExt(path) {
-			logVerbose(grey(" - ", relPath, " (no code)"))
+		if args.AllValidText && !isValidUtf8(path) {
```
Comment (gwillem) [inline corediff.go:197]: duplicate call to isValidUtf8, could store result if necessary

File: corediff.go [hunk 4]
```diff
@@ -186,9 +194,14 @@ func addPath(root string, db hashDB, args *baseArgs) {
...
+			logVerbose(grey(" - ", relPath, " (invalid utf8)"))
 			return nil
+		} else {
+			if !hasValidExt(path) {
```
Comment (gwillem) [inline corediff.go:201]: `} else if {`

Also, switch order here and put cheapest call first

**Comment by gwillem** [comment]
Awesome work Kevo! My main feedback is that the code gets more complex because of the AllValidText argument, while that option is only required for this one-off test against our malware corpus. Or do you see valid use case during DFIR cases?

(Alternatively you could have copied the backend corpus, removed binary files, and forced the rest to a .php extension so it would be included in the scan?)

**Comment by gwillem** [comment]
Great job on the stats also!

File: corediff.go (hunk 3 above)
Comment (SKevo18) [reply corediff.go:197]: The other call is related to `args.Add`, and shouldn't be run twice if we scan or add the files AFAIK, if that's what you mean?

**Comment by SKevo18** [comment]
> My main feedback is that the code gets more complex because of the AllValidText argument, while that option is only required for this one-off test against our malware corpus. Or do you see valid use case during DFIR cases?

No other valid use case other than when the malware doesn't use a file extension corediff recognizes. In that case, it might sometimes be handy to scan all text-like files (including crontabs, for example). It is optional, meaning that we can use it only when we want to so it doesn't slow down normal scan. But I agree that we probably won't be using it as much.

> (Alternatively you could have copied the backend corpus, removed binary files, and forced the rest to a .php extension so it would be included in the scan?)

Indeed, but I don't mind adding it. I think it's okay because we can decide whether we want to keep it or not - and if not, then at least it helped during the corpus analysis. The work needed to copy the corpus and filter binary files properly is comparable to adding the flag IMO. Also, there would not be a potential addition - the flag - if I just filtered binary files instead. I like to experiment

**Comment by gwillem** [comment]
> The work needed to copy the corpus and filter binary files properly is comparable to adding the flag IMO.

If they require the same effort, but one option creates future cost (added complexity) then the other is preferable?

File: corediff.go (hunk 3 above)
Comment (gwillem) [reply corediff.go]: you're right, i overlooked that!

**Review by gwillem — APPROVED** [review]

**Comment by SKevo18** [comment]
> If they require the same effort, but one option creates future cost (added complexity) then the simpler is preferable?

Yes, I agree that the additional complexity is one of the downsides. But I am not sure if that con outweighs the pro of potentially detecting malware outside of the allowed file extensions... Do you think we should keep the flag or remove it to keep the code simpler and easier to maintain?

**Comment by gwillem** [comment]
Up to you!

**Pull request merged**